diff --git a/core/modules/field_ui/src/Form/EntityViewDisplayEditForm.php b/core/modules/field_ui/src/Form/EntityViewDisplayEditForm.php
index a188af5d92d..6052fe5e3f5 100644
--- a/core/modules/field_ui/src/Form/EntityViewDisplayEditForm.php
+++ b/core/modules/field_ui/src/Form/EntityViewDisplayEditForm.php
@@ -162,13 +162,15 @@ protected function thirdPartySettingsForm(PluginSettingsInterface $plugin, Field
     $this->moduleHandler->invokeAllWith(
       'field_formatter_third_party_settings_form',
       function (callable $hook, string $module) use (&$settings_form, &$plugin, &$field_definition, &$form, &$form_state) {
-        $settings_form[$module] = ($settings_form[$module] ?? []) + ($hook(
+        $hook_result = $hook(
           $plugin,
           $field_definition,
           $this->entity->getMode(),
           $form,
-          $form_state,
-        )) ?? [];
+          $form_state
+        ) ?? [];
+        
+        $settings_form[$module] = ($settings_form[$module] ?? []) + $hook_result;
       }
     );
     return $settings_form;
